supernode/daemon/mgr/progress/progress_util.go

```diff
@@ -0,0 +1,241 @@
+			return err
+		}
+	}
+	if !errorType.IsDataNotFound(err) {
+		return err
+	}
+
+	return nil
+}
+
+// processPeerFailInfo adds one to the count of errors
+// when srcCID failed to download a piece from dstCID.
+// And add the dstCID to the blacklist of the srcCID.
+func (pm *ProgressManager) processPeerFailInfo(srcCID, dstCID string) (err error) {
+	if !needDoPeerInfo(srcCID, dstCID) {
+		return nil
+	}
+
+	// update black List
+	blackList, err := pm.clientBlackInfo.GetAsMap(srcCID)
+	if err != nil && errorType.IsDataNotFound(err) {
+		blackList = cutil.NewSyncMap()
+		if err := pm.clientBlackInfo.Add(srcCID, blackList); err != nil {
+			return err
+		}
+	}
+	if _, err := blackList.Get(dstCID); err != nil &&
+		errorType.IsDataNotFound(err) {
+		if err := blackList.Add(dstCID, true); err != nil {
+			return err
+		}
+	}
+
+	// update clientErrorInfo
+	clientErrorCount, err := pm.clientErrorInfo.GetAsInt(srcCID)
+	if err == nil {
+		if err := pm.clientErrorInfo.Add(srcCID, clientErrorCount+1); err != nil {
+			return err
+		}
+	}
+	if !errorType.IsDataNotFound(err) {
+		return err
+	}
+	if err := pm.clientErrorInfo.Add(srcCID, clientErrorCount+1); err != nil {
+		return err
+	}
+
+	// update serviceErrorInfo
+	serviceErrorCount, err := pm.serviceErrorInfo.GetAsInt(dstCID)
+	if err == nil {
+		return pm.serviceErrorInfo.Add(dstCID, serviceErrorCount+1)
+	}
+
+	if !errorType.IsDataNotFound(err) {
+		return err
+	}
+	return pm.serviceErrorInfo.Add(dstCID, 1)
+}
+
+// needDoPeerInfo returns whether we should update the peer related info.
+// It returns false when the cid is empty or represents a supernode.
+func needDoPeerInfo(srcCID, dstCID string) bool {
+	if cutil.IsEmptyStr(srcCID) || cutil.IsEmptyStr(srcCID) ||
+		isSupernode(srcCID) || isSupernode(dstCID) {
+		return false
+	}
+	return true
+}
+
+// isSupernode returns whether the clientID represents supernode.
+// TODO: implement it.
+var isSupernode = func(clientID string) bool {
+	return strings.HasPrefix(clientID, config.SuperNodeCIdPrefix)
+}
+
+// generatePieceProgressKey returns a string as the key of PieceProgress.
+func generatePieceProgressKey(taskID string, pieceNum int) (string, error) {
+	if cutil.IsEmptyStr(taskID) || pieceNum < 0 {
+		return "", errors.Wrapf(errorType.ErrInvalidValue,
+			"failed to make piece progress key with taskID: %s and pieceNum: %d", taskID, pieceNum)
+	}
+	return fmt.Sprintf("%d@%s", pieceNum, taskID), nil
+}
+
+// getAsPieceState returns result as *pieceState.
+// The ErrConvertFailed error will be returned if the assertion fails.
+func getAsPieceState(mmap *cutil.SyncMap, key string) (*pieceState, error) {
+	v, err := mmap.Get(key)
+	if err != nil {
+		return nil, errors.Wrapf(err, "key: %s", key)
+	}
+
+	if value, ok := v.(*pieceState); ok {
+		return value, nil
+	}
+	return nil, errors.Wrapf(errorType.ErrConvertFailed, "key %s: %v", key, v)
+}
+
+func updatePieceBitset(pieceBitSet *bitset.BitSet, pieceNum, pieceStatus int) bool {
```

Comment:
Use `BitSet` instead of `Bitset` to consistent with the other places.

Reply:
Done. 😄